projects/driver_controls_pedal/src/main.c

```diff
@@ -0,0 +1,144 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "bps_indicator.h"
+#include "calib.h"
+#include "can_msg_defs.h"
+#include "direction_indicator.h"
+#include "event_queue.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "heartbeat_rx.h"
+#include "i2c.h"
+#include "interrupt.h"
+#include "log.h"
+#include "pc_input_event.h"
+#include "soft_timer.h"
+#include "throttle.h"
+
+#include "brake_signal.h"
+#include "event_arbiter.h"
+#include "mech_brake_fsm.h"
+#include "pedal_fsm.h"
+#include "pedal_output.h"
+#include "power_state_indicator.h"
+
+#include "can.h"
+#include "crc32.h"
+#include "flash.h"
+#include "pc_calib.h"
+#include "pc_cfg.h"
+
+typedef StatusCode (*PedalControlsFsmInitFn)(Fsm *fsm, EventArbiterStorage *storage);
+
+typedef enum {
+  PEDAL_CONTROLS_FSM_PEDAL = 0,
+  PEDAL_CONTROLS_FSM_MECH_BRAKE,
+  NUM_PEDAL_CONTROLS_FSMS,
+} PedalControlsFsm;
+
+static PcCalibBlob s_calib_blob;
+static Ads1015Storage s_pedal_ads1015;
+static MechBrakeStorage s_mech_brake;
+static EventArbiterStorage s_event_arbiter;
+static Fsm s_fsms[NUM_PEDAL_CONTROLS_FSMS];
+static CanStorage s_can;
+static Fsm s_fsms[NUM_PEDAL_CONTROLS_FSMS];
+static HeartbeatRxHandlerStorage s_powertrain_heartbeat;
+
+int main(void) {
+  gpio_init();
+  interrupt_init();
+  gpio_it_init();
+  soft_timer_init();
+  event_queue_init();
+  crc32_init();
+  flash_init();
+
+  const CanSettings can_settings = {
+    .device_id = PC_CFG_CAN_DEVICE_ID,
+    .bitrate = PC_CFG_CAN_BITRATE,
+    .rx_event = INPUT_EVENT_PEDAL_CAN_RX,
+    .tx_event = INPUT_EVENT_PEDAL_CAN_TX,
+    .fault_event = INPUT_EVENT_PEDAL_CAN_FAULT,
+    .tx = PC_CFG_CAN_TX,
+    .rx = PC_CFG_CAN_RX,
+    .loopback = false,
+  };
+
+  can_init(&s_can, &can_settings);
+  can_add_filter(SYSTEM_CAN_MESSAGE_DRIVE_OUTPUT);
+  can_add_filter(SYSTEM_CAN_MESSAGE_POWER_STATE);
+
+  const I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,    //
+    .sda = PC_CFG_I2C_BUS_SDA,  //
+    .scl = PC_CFG_I2C_BUS_SCL,  //
+  };
+
+  i2c_init(PC_CFG_I2C_BUS_PORT, &i2c_settings);
+
+  // Power state indicator
+  power_state_indicator_init();
+
+  // Direction state indicator
+  direction_indicator_init();
+
+  // BPS heartbeat
+  bps_indicator_init();
+
+  // Powertrain heartbeat
+  heartbeat_rx_register_handler(&s_powertrain_heartbeat, SYSTEM_CAN_MESSAGE_POWERTRAIN_HEARTBEAT,
+                                heartbeat_rx_auto_ack_handler, NULL);
+
+  GpioAddress pedal_ads1015_ready = PC_CFG_PEDAL_ADC_RDY_PIN;
+  ads1015_init(&s_pedal_ads1015, PC_CFG_I2C_BUS_PORT, PC_CFG_PEDAL_ADC_ADDR, &pedal_ads1015_ready);
+  PcCalibBlob *pc_calib_blob = calib_blob();
+  throttle_init(throttle_global(), &pc_calib_blob->throttle_calib, &s_pedal_ads1015);
+
+  const MechBrakeSettings mech_brake_settings = {
+    .ads1015 = &s_pedal_ads1015,
+    .brake_pressed_threshold_percentage = 55,
+    .brake_unpressed_threshold_percentage = 45,
```

Comment:
Where are these numbers from?

Reply:
The numbers were chosen by pressing/unpressing the brake to figure out what a good threshold percentage would be. They were chosen last year before the race by one of the previous leads I believe.

Reply:
Weren't those based on Pedal Board Rev 2 with the magnetic sensor? Do these still work based on the potentiometer that we chose for Rev 3? If so, where does 45/55% correspond to roughly in terms of brake travel?

Reply:
The mech brake is going to be re-calibrated when the pedal board is mounted so these values may or may not change based on the range of brake travel. For now, these serve as the default values.

Reply:
Sure, but what's the input range that we can expect from the brake pedal from not pressed to fully pressed with the new potentiometer? So like what's the range of signals that you're trying to calibrate for?

Reply:
This setting adds hysteresis to the post-calibration brake percentage. It doesn't matter which sensor you use.